pkg/controllers/gateway/gateway_controller.go

```diff
@@ -351,6 +375,99 @@ func selectClusters(gateway gatewayv1beta1.Gateway) []string {
 	return []string{}
 }
 
+func (r *GatewayReconciler) DNSDeletion(ctx context.Context, gateway gatewayv1beta1.Gateway) error {
+	trafficAccessor := traffic.NewGateway(&gateway)
+	allHosts := trafficAccessor.GetHosts()
+
+	for _, host := range allHosts {
+		managedZone, subDomain, err := r.Host.GetManagedZoneForHost(ctx, host, trafficAccessor)
+		if err != nil {
+			return err
+		}
+		if managedZone == nil {
+			log.Log.Info("Managed zone not found cannot get DNS record")
+			return nil
+		}
+
+		log.Log.Info("getting DNS record associated with deleting gateway")
+		dnsRecord, err := r.Host.GetDNSRecord(ctx, subDomain, managedZone)
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return err
+		}
+		if dnsRecord == nil {
+			log.Log.Info("dns record not found")
+			continue
+		}
+
+		log.Log.Info("deleting DNS record ", "DNSRecord:", dnsRecord.Name)
+		err = r.Delete(ctx, dnsRecord)
+		log.Log.Info("DNS record deleted ", "DNSRecord:", dnsRecord.Name)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *GatewayReconciler) cleanupCertificates(ctx context.Context, gateway *gatewayv1beta1.Gateway) error {
+	var hostsToRemove []string
+	var allGateways gatewayv1beta1.GatewayList
+
+	// get names of hosts for traffic object being deleted
+	for _, listener := range gateway.Spec.Listeners {
```

Comment:
I think we could check the protocol of the listener here 

```listener.Protocol``` should be HTTPS

Reply:
Would there be a case when based on the protocol we would decide to keep the hostname? Right now it tries to get rid of all of them regardless the protocol